gloo/allreduce.cc

```diff
@@ -216,13 +216,21 @@ void ring(
   //
   const size_t numSegments = roundUp(
       std::max(
-          (totalBytes + (segmentBytes - 1)) / segmentBytes,
+          (totalBytes + (tmpSegmentBytes - 1)) / tmpSegmentBytes,
```

Comment:
I am trying to think if there is a more straightforward way to calculate `numSegments` and `segmentBytes`. Does the following looks reasonable? This avoids going through the `tmpSegmentBytes`.

```c++
const size_t maxSegmentSize = std::max(
    opts.elementSize * (opts.maxSegmentSize / opts.elementSize),
    opts.elementSize);

const size_t numSegments = std::max(
    context->size * 2,
    (totalBytes + (maxSegmentSize - 1)) / maxSegmentSize
);

const size_t segmentBytes = roundUp(
    (totalBytes + numSegments - 1) / numSegments,
    opts.elementSize);

```

Reply:
There probably is, thanks for the suggested snippet. Let me see if it works and update the PR.